keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -557,278 +532,231 @@ reconstructed by the driver. This has no effect on already allocated claims
-`ResourceSlice`. The implementations of structured models must handle
-scenarios where more resources are allocated than available. The kubelet plugin
-of a DRA driver ***must*** double-check that the allocated resources are still
+**Note:** If a driver needs to remove a device or change its attributes,
+then there is a risk that a claim gets allocated based on the old
+`ResourceSlice`. The scheduler must handle
+scenarios where more devices are allocated than available. The kubelet plugin
+of a DRA driver ***must*** double-check that the allocated devices are still
 available when NodePrepareResource is called. If not, the pod cannot start until
-the resource comes back. Treating this as a fatal error during pod admission
-would allow us to delete the pod and trying again with a new one.
-
-### Using structured parameters
-
-The following is an example CRD which the developer of the
-`cards.dra.example.com` DRA driver might define as a valid claim parameters
-object for requesting access to its GPUs:
+the device comes back. Checking it at admission time and treating this as a fatal error
+would allow us to delete the pod and trying again with a new one, but is not done
+at the moment because admission checks cannot be retried if a check finds
+a transient problem.
+
+#### Partitionable devices
+
+In addition to devices, a `ResourceSlice` can also embed a list of
+`SharedCapacity` objects. Each `SharedCapacity` object represents some amount
+of shared "capacity" that can be consumed by one or more devices listed in the
+slice. When listing such devices, the sum of the capacity consumed across all
+devices *may* exceed the total amount available in the `SharedCapacity` object.
+This allows one, for example, to provide a way of logically partition a device
+into a set of overlapping sub-devices, each of which "could" be allocated by a
+scheduler (just not at the same time).
+
+As such, scheduler support will need to be added to track the set of
+`SharedCapacity` objects provided by each `ResourceSlice` as well as perform
+the following steps to decide if a given device is a candidate for allocation
+or not:
+
+1. Look at the set of `SharedCapacityConsumed` objects referenced by the device
+1. For each `SharedCapacityConsumed` object, look and see how much capacity is still available in the `SharedCapacity` object it is tracking with the same name
+1. If enough capacity is still available in the `SharedCapacity` object to satisfy the device's consumption, continue to consider it for allocation
+1. If not enough capacity is available, move on to the next device
+1. Upon deciding to allocate a device, subtract all of the capacity in its `SharedCapacityConsumed` objects from the corresponding `SharedCapacity` objects being tracked
+1. Upon freeing a device, add all of the capacity in its `SharedCapacityConsumed` objects back into the corresponding `SharedCapacity` objects being tracked
```

Comment:
@alculquicondor I just added this text today, which has some (minor) scheduler implications.

Reply:
[Looked good](https://github.com/kubernetes/enhancements/pull/4709#pullrequestreview-2113871842) to him, resolving.